src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -186,30 +378,46 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                          const void *val)
 {
     int mpl_err = MPL_SUCCESS;
+    int direction;
     gavl_tree_node_s *node_ptr;
+    gavl_tree_node_s *cur_node;
     gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
 
     node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
     if (node_ptr == NULL) {
         mpl_err = MPL_ERR_SHM_NOMEM;
         goto fn_fail;
     }
+    node_ptr->addr = (uintptr_t) addr;
+    node_ptr->len = len;
+
+    node_ptr->subnode_list =
+        (gavl_tree_subnode_s *) MPL_malloc(sizeof(gavl_tree_subnode_s), MPL_MEM_OTHER);
+    if (node_ptr->subnode_list == NULL) {
+        mpl_err = MPL_ERR_SHM_NOMEM;
+        goto fn_fail;
+    }
+    node_ptr->subnode_list->addr = (uintptr_t) addr;
+    node_ptr->subnode_list->len = len;
+    node_ptr->subnode_list->val = val;
+    node_ptr->subnode_list->next = NULL;
+
+    DECLARE_STACK(gavl_tree_node_s *, node_stack);
+
+  restart:
```

Comment:
Moreover, can we create `subnode_list` only when you insert subset buffers into the node? The concern is that, for most tree nodes which do not contain any subset buffer, we have to always perform an additional pointer deference to access node_ptr->subnode_list[0].

Reply:
> I don't understand why you always create a new node before searching if there is already a matching buffer in the tree. Do you assume that the caller always call `MPL_gavl_tree_search` before insert?

search and insert function are separate. I don't assume `MPL_gavl_tree_search` is called before insert. Even if there is a matching in the buffer, we can still insert the new node which won't affect correctness.

> Moreover, can we create `subnode_list` only when you insert subset buffers into the node? The concern is that, for most tree nodes which do not contain any subset buffer, we have to always perform an additional pointer deference to access node_ptr->subnode_list[0].

we can, but I don't recommend to do it. 
The reason is 
1. this optimization will indeed save a pointer deference, but we are using it for large message transfer, so I think one deference won't affect performance too much. However, if we do this optimization, code will become much more complex.

Reply:
Inserting a matched buffer may not affect correctness but it seems unnecessary. I feel that it makes the subnode logic overcomplicated. Please see my questions at `MPL_gavl_tree_delete `.

The AVL tree might be used for small message too. For example, I am implementing native RMA for dynamic window now. I need store all memory regions registered at win_attach on each process. For any RMA operation (even if it is small) that is issued via this window, I need search the AVL tree to get the corresponding remote MR key.

Actually, for all of our use cases, can we simply delete the tree nodes when inserting a new region whose buffer is a superset of the existing ones? E.g., existing node is [10,20], [50,60], when inserting [0,100], both [10,20] and [50,60] will be deleted and [0, 100] will be inserted. If we do this, do we still have the mismatching issue?

Reply:
> The AVL tree might be used for small message too. For example, I am implementing native RMA for dynamic window now. I need store all memory regions registered at win_attach on each process. For any RMA operation (even if it is small) that is issued via this window, I need search the AVL tree to get the corresponding remote MR key.

I am not sure what is dynamic window, but for `win_attach`, I think XPMEM buffer register happens in this function and then we can store the attached addr into win object. There will be no search/insert during RMA operations. I am not sure what MR key is, maybe I miss something. Could you please correct me if I am wrong?

> Actually, for all of our use cases, can we simply delete the tree nodes when inserting a new region whose buffer is a superset of the existing ones? E.g., existing node is [10,20], [50,60], when inserting [0,100], both [10,20] and [50,60] will be deleted and [0, 100] will be inserted. If we do this, do we still have the mismatching issue?

This is an incorrect design. The subnode design actually tries to solve this issue. In network buffer registration, if we first register [10, 20], and sending is still ongoing (e.g. isend); then if we call another isend, try to register [0, 100] and delete [10, 20], at this time, the previous sending is ongoing but the buffer is invalidated. It will cause an error.

Reply:
For dynamic window, a process can attach multiple memory regions to the window. For every attached region, if we want to use hardware RDMA, we need register the region to network driver. `MR key` is the ID returned from network adapter at memory registration. When issuing an RMA via HW RDMA, we need check the remote region associated with the `target_buffer`, and get its remote `MR key` to issue the RDMA. In summary, we need `cache` the remote regions (MR key, addr, size) attached by each process, and `search` the it by using `target_buf` and `target_sz` when issuing an RMA. At `win_detach` or `win_free`, the cached regions need to be `deleted`.

I see the issue if we want to delete a previously cached buffer. However, if we only delete the tree node but not actually deregister the buffer, should it be fine?

Besides, it is illegal to attach overlapped regions in `win_attach`. In GPU IPC, I think it is the same that we never need to deal with overlapping issue since we always cache the entire buffer returned from cudaMalloc. In the above cases, the subnode code seems overcomplicated and introduces unnecessary overhead (e.g., pointer deference at search)

Reply:
> For dynamic window, a process can attach multiple memory regions to the window. For every attached region, if we want to use hardware RDMA, we need register the region to network driver. `MR key` is the ID returned from network adapter at memory registration. When issuing an RMA via HW RDMA, we need check the remote region associated with the `target_buffer`, and get its remote `MR key` to issue the RDMA. In summary, we need `cache` the remote regions (MR key, addr, size) attached by each process, and `search` the it by using `target_buf` and `target_sz` when issuing an RMA. At `win_detach` or `win_free`, the cached regions need to be `deleted`.

I see. but I am not sure whether a pointer dereference will affect a lot in network communication when the message size is small. Since saving this one pointer dereference will make code more complex, if it can affect around 20% or more performance, I think it is worth changing the code.

> I see the issue if we want to delete a previously cached buffer. However, if we only delete the tree node but not actually deregister the buffer, should it be fine?

when tree node is deleted, we don't deregister the buffer since we need it to merge with the input node and reinsert the merged node into tree in next round. Insert function should not delete any buffer.

> Besides, it is illegal to attach overlapped regions in `win_attach`. In GPU IPC, I think it is the same that we never need to deal with overlapping issue since we always cache the entire buffer returned from cudaMalloc. In the above cases, the subnode code seems overcomplicated and introduces unnecessary overhead (e.g., pointer deference at search)

you mean network registration buffer can never overlap (I think in InfiniBand, it might be possible to overlap the buffer)? I agree with GPU, but gavl tree is not only used by GPU. We need to consider the cases where buffer may overlap.

Reply:
> introduces unnecessary overhead (e.g., pointer deference at search)

I have to agree with Kaiming on this.  We are talking about large messages here.  I'll be shocked to if there is any noticeable difference from a pointer dereference in such cases.  I'd vote for maintainability unless there's a demonstrable performance impact of the dereference.